core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,137 @@
+      size_t maxWait,
+      std::shared_ptr<FullNodeApi> api,
+      std::shared_ptr<libp2p::protocol::Scheduler> scheduler,
+      Address &miner_address) {
+    struct make_unique_enabler : public PreCommitBatcherImpl {
+      make_unique_enabler(size_t MaxTime, std::shared_ptr<FullNodeApi> api)
+          : PreCommitBatcherImpl(MaxTime, std::move(api)){};
+    };
+
+    std::shared_ptr<PreCommitBatcherImpl> batcher =
+        std::make_unique<make_unique_enabler>(maxWait, std::move(api));
+
+    batcher->miner_address_ = miner_address;
+    batcher->maxDelay = maxWait;
+
+    batcher->cutoffStart = std::chrono::system_clock::now();
+    batcher->closestCutoff = maxWait;
+    batcher->handle_ = scheduler->schedule(maxWait, [=]() {
+      auto maybe_send = batcher->sendBatch();
+      batcher->handle_.reschedule(maxWait);  // TODO: maybe in gsl::finally
+    });
+    return batcher;
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::sendBatch() {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+
+    OUTCOME_TRY(head, api_->ChainHead());
+    OUTCOME_TRY(minfo, api_->StateMinerInfo(miner_address_, head->key));
+
+    PreCommitBatch::Params params = {};
+    for (auto &data : batchStorage) {
+      mutualDeposit += data.second.deposit;
+      params.sectors.push_back(data.second.pci);
+    }
+    // TODO: goodFunds = mutualDeposit + MaxFee; -  for checking payable
+    if (params.sectors.size() != 0) {
+      OUTCOME_TRY(encodedParams, codec::cbor::encode(params));
+
+      auto maybe_signed = api_->MpoolPushMessage(
+          vm::message::UnsignedMessage(
+              miner_address_,
+              minfo.worker,  // TODO: handle worker
+              0,
+              mutualDeposit,
+              {},
+              {},
+              vm::actor::builtin::v0::miner::PreCommitSector::Number,
+              MethodParams{encodedParams}),
+          kPushNoSpec);
+
+      if (maybe_signed.has_error()) {
+        return maybe_signed.error(); //TODO: maybe logs
+      }
+
+      mutualDeposit = 0;
+      batchStorage.clear();
+    }
+    cutoffStart = std::chrono::system_clock::now();
+    return outcome::success();
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::forceSend() {
+    auto error = sendBatch();
+    assert(not error.has_error());
+    handle_.reschedule(maxDelay);
+    return outcome::success();
+  }
+
+  void PreCommitBatcherImpl::getPreCommitCutoff(ChainEpoch curEpoch,
+                                                const SectorInfo &si) {
+    primitives::ChainEpoch cutoffEpoch =
+        si.ticket_epoch + static_cast<int64_t>(fc::kEpochsInDay + 600);
+    ChainEpoch startEpoch{};
+    for (auto p : si.pieces) {
+      if (!p.deal_info) {
+        continue;
+      }
+      startEpoch = p.deal_info->deal_schedule.start_epoch;
+      if (startEpoch < cutoffEpoch) {
+        cutoffEpoch = startEpoch;
+      }
+    }
+    if (cutoffEpoch <= curEpoch) {
+      auto maybe_error = forceSend();
```

Comment:
Error is not checked. May be it is ok since `forceSend()` always returns `success`, but I think something should be changed.

Reply:
In such cases use
```suggestion
      OUTCOME_TRY(forceSend());
```
And change `void getPreCommitCutoff` to `outcome::result<void> getPreCommitCutoff`

Reply:
Moreover, I think that "get" is not suitable prefix for the method returns void

Reply:
fixed